File: archive/cdk-ts/Pr921/lib/webapp-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as elbv2 from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import * as autoscaling from 'aws-cdk-lib/aws-autoscaling';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

interface WebAppStackProps extends cdk.StackProps {
  environmentSuffix: string;
  vpc: ec2.Vpc;
  regionName: string;
}

export class WebAppStack extends cdk.Stack {
  public readonly loadBalancer: elbv2.ApplicationLoadBalancer;

  constructor(scope: Construct, id: string, props: WebAppStackProps) {
    super(scope, id, props);

    // Create security group for ALB
    const albSecurityGroup = new ec2.SecurityGroup(
      this,
      `ALBSG-${props.regionName}`,
      {
        vpc: props.vpc,
        description: 'Security group for Application Load Balancer',
        allowAllOutbound: true,
      }
    );

    albSecurityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80),
      'Allow HTTP traffic from internet'
    );

    albSecurityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(443),
      'Allow HTTPS traffic from internet'
    );

    // Create Application Load Balancer
    this.loadBalancer = new elbv2.ApplicationLoadBalancer(
      this,
      `WebAppALB-${props.regionName}`,
      {
        vpc: props.vpc,
        internetFacing: true,
        securityGroup: albSecurityGroup,
        loadBalancerName: `alb-${props.regionName.substring(0, 3)}-${props.environmentSuffix.substring(0, 20)}`,
      }
    );

    // Create security group for EC2 instances
    const ec2SecurityGroup = new ec2.SecurityGroup(
      this,
      `EC2SG-${props.regionName}`,
      {
        vpc: props.vpc,
        description: 'Security group for EC2 instances',
        allowAllOutbound: true,
      }
    );

    ec2SecurityGroup.addIngressRule(
      albSecurityGroup,
      ec2.Port.tcp(80),
      'Allow HTTP traffic from ALB'
    );

    // Create IAM role for EC2 instances
    const ec2Role = new iam.Role(this, `EC2Role-${props.regionName}`, {
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          'AmazonSSMManagedInstanceCore'
        ),
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          'CloudWatchAgentServerPolicy'
        ),
      ],
    });

    // User data script for web server setup
    const userData = ec2.UserData.forLinux();
    userData.addCommands(
      'yum update -y',
      'yum install -y httpd',
      'systemctl start httpd',
      'systemctl enable httpd',
      'echo "<html><body><h1>Web Application - Region: ${AWS::Region}</h1><p>Instance ID: $(curl -s http://169.254.169.254/latest/meta-data/instance-id)</p></body></html>" > /var/www/html/index.html'
    );

    // Create Launch Template
    const launchTemplate = new ec2.LaunchTemplate(
      this,
      `LaunchTemplate-${props.regionName}`,
      {
        launchTemplateName: `lt-${props.regionName.substring(0, 3)}-${props.environmentSuffix}`,
        instanceType: ec2.InstanceType.of(
          ec2.InstanceClass.T3,
          ec2.InstanceSize.MICRO
        ),
        machineImage: ec2.MachineImage.latestAmazonLinux2(),
        securityGroup: ec2SecurityGroup,
        role: ec2Role,
        userData: userData,
        detailedMonitoring: true,
      }
    );

    // Create Auto Scaling Group
    const autoScalingGroup = new autoscaling.AutoScalingGroup(
      this,
      `ASG-${props.regionName}`,
      {
        vpc: props.vpc,
        launchTemplate: launchTemplate,
        minCapacity: 2,
        maxCapacity: 10,
        desiredCapacity: 2,
        vpcSubnets: {
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        },
        healthCheck: autoscaling.HealthCheck.elb({
          grace: cdk.Duration.minutes(5),
        }),
        autoScalingGroupName: `asg-${props.regionName.substring(0, 3)}-${props.environmentSuffix}`,
      }
    );

    // Create target group
    const targetGroup = new elbv2.ApplicationTargetGroup(
      this,
      `TargetGroup-${props.regionName}`,
      {
        vpc: props.vpc,
        port: 80,
        protocol: elbv2.ApplicationProtocol.HTTP,
        targetType: elbv2.TargetType.INSTANCE,
        healthCheck: {
          enabled: true,
          healthyHttpCodes: '200',
          interval: cdk.Duration.seconds(30),
          timeout: cdk.Duration.seconds(10),
          unhealthyThresholdCount: 5,
          path: '/',
        },
        targetGroupName: `tg-${props.regionName.substring(0, 3)}-${props.environmentSuffix.substring(0, 20)}`,
      }
    );

    // Attach ASG to target group
    autoScalingGroup.attachToApplicationTargetGroup(targetGroup);

    // Add listener to ALB
    this.loadBalancer.addListener(`Listener-${props.regionName}`, {
      port: 80,
      protocol: elbv2.ApplicationProtocol.HTTP,
      defaultTargetGroups: [targetGroup],
    });

    // Add scaling policies
    autoScalingGroup.scaleOnCpuUtilization(`CPUScaling-${props.regionName}`, {
      targetUtilizationPercent: 70,
      cooldown: cdk.Duration.minutes(5),
    });

    autoScalingGroup.scaleOnRequestCount(`RequestScaling-${props.regionName}`, {
      targetRequestsPerMinute: 1000,
      cooldown: cdk.Duration.minutes(5),
    });

    // Export ALB DNS name for Route 53
    new cdk.CfnOutput(this, `LoadBalancerDNS-${props.regionName}`, {
      value: this.loadBalancer.loadBalancerDnsName,
      exportName: `ALB-DNS-${props.regionName}-${props.environmentSuffix}`,
    });

    new cdk.CfnOutput(this, `LoadBalancerHostedZoneId-${props.regionName}`, {
      value: this.loadBalancer.loadBalancerCanonicalHostedZoneId,
      exportName: `ALB-HZ-${props.regionName}-${props.environmentSuffix}`,
    });
  }
}

File: archive/cdk-java/Pr2258/lib/src/main/java/app/WebAppStack.java
package app;

import software.amazon.awscdk.CfnOutput;
import software.amazon.awscdk.CfnOutputProps;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.RemovalPolicy;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.Tags;
import software.amazon.awscdk.services.autoscaling.AutoScalingGroup;
import software.amazon.awscdk.services.autoscaling.HealthCheck;
import software.amazon.awscdk.services.cloudwatch.Metric;
import software.amazon.awscdk.services.certificatemanager.Certificate;
import software.amazon.awscdk.services.certificatemanager.CertificateValidation;
import software.amazon.awscdk.services.certificatemanager.ValidationMethod;
import software.amazon.awscdk.services.ec2.*;
import software.amazon.awscdk.services.elasticloadbalancingv2.*;
import software.amazon.awscdk.services.elasticloadbalancingv2.targets.InstanceIdTarget;
import software.amazon.awscdk.services.iam.ManagedPolicy;
import software.amazon.awscdk.services.iam.Role;
import software.amazon.awscdk.services.iam.ServicePrincipal;
import software.amazon.awscdk.services.iam.InstanceProfile;
import software.amazon.awscdk.services.s3.Bucket;
import software.amazon.awscdk.services.s3.BucketEncryption;
import software.amazon.awscdk.services.s3.LifecycleRule;
import software.amazon.awscdk.services.s3.StorageClass;
import software.amazon.awscdk.services.s3.Transition;
import software.constructs.Construct;
import software.amazon.awscdk.Stack;

import java.util.List;
import java.util.Map;

/**
 * Stack for deploying a highly available web application with Auto Scaling,
 * Application Load Balancer, and S3 logging infrastructure.
 */
public class WebAppStack extends Stack {
    
    private final Vpc vpc;
    private final ApplicationLoadBalancer alb;
    private final AutoScalingGroup autoScalingGroup;
    private final Bucket logsBucket;
    private final String environmentSuffix;

    public WebAppStack(final Construct scope, final String id, final WebAppStackProps props) {
        super(scope, id, props.getStackProps());
        
        this.environmentSuffix = props.getEnvironmentSuffix();
        
        // Create VPC with public and private subnets
        this.vpc = createVpc();
        
        // Create S3 bucket for logs
        this.logsBucket = createLogsBucket();
        
        // Create security groups
        SecurityGroup albSecurityGroup = createAlbSecurityGroup();
        SecurityGroup instanceSecurityGroup = createInstanceSecurityGroup(albSecurityGroup);
        
        // Create IAM role for EC2 instances
        Role instanceRole = createInstanceRole();
        
        // Create launch template
        LaunchTemplate launchTemplate = createLaunchTemplate(instanceSecurityGroup, instanceRole);
        
        // Create Auto Scaling Group
        this.autoScalingGroup = createAutoScalingGroup(launchTemplate);
        
        // Create Application Load Balancer
        this.alb = createApplicationLoadBalancer(albSecurityGroup);
        
        // Create certificate and configure HTTPS
        configureSslAndTargetGroups();
        
        // Apply tags to all resources
        applyTags();
        
        // Create outputs
        createOutputs();
    }
    
    private Vpc createVpc() {
        return Vpc.Builder.create(this, "WebAppVpc" + environmentSuffix)
            .maxAzs(3)
            .cidr("10.0.0.0/16")
            .natGateways(2) // For high availability
            .subnetConfiguration(List.of(
                SubnetConfiguration.builder()
                    .name("Public")
                    .subnetType(SubnetType.PUBLIC)
                    .cidrMask(24)
                    .build(),
                SubnetConfiguration.builder()
                    .name("Private")
                    .subnetType(SubnetType.PRIVATE_WITH_EGRESS)
                    .cidrMask(24)
                    .build()
            ))
            .build();
    }
    
    private Bucket createLogsBucket() {
        return Bucket.Builder.create(this, "WebAppLogsBucket" + environmentSuffix)
            .bucketName("webapp-logs-" + environmentSuffix.toLowerCase() + "-" + this.getRegion())
            .versioned(true)
            .encryption(BucketEncryption.S3_MANAGED)
            .lifecycleRules(List.of(
                LifecycleRule.builder()
                    .id("LogsLifecycle")
                    .enabled(true)
                    .transitions(List.of(
                        Transition.builder()
                            .storageClass(StorageClass.GLACIER)
                            .transitionAfter(Duration.days(30))
                            .build()
                    ))
                    .expiration(Duration.days(365))
                    .build()
            ))
            .removalPolicy(RemovalPolicy.DESTROY)
            .autoDeleteObjects(true)
            .build();
    }
    
    private SecurityGroup createAlbSecurityGroup() {
        SecurityGroup sg = SecurityGroup.Builder.create(this, "AlbSecurityGroup" + environmentSuffix)
            .vpc(vpc)
            .description("Security group for Application Load Balancer")
            .allowAllOutbound(true)
            .build();
            
        // Allow HTTP and HTTPS traffic from anywhere
        sg.addIngressRule(Peer.anyIpv4(), Port.tcp(80), "Allow HTTP traffic");
        sg.addIngressRule(Peer.anyIpv4(), Port.tcp(443), "Allow HTTPS traffic");
        
        return sg;
    }
    
    private SecurityGroup createInstanceSecurityGroup(SecurityGroup albSecurityGroup) {
        SecurityGroup sg = SecurityGroup.Builder.create(this, "InstanceSecurityGroup" + environmentSuffix)
            .vpc(vpc)
            .description("Security group for EC2 instances")
            .allowAllOutbound(true)
            .build();
            
        // Allow traffic from ALB only
        sg.addIngressRule(Peer.securityGroupId(albSecurityGroup.getSecurityGroupId()), 
                         Port.tcp(80), "Allow HTTP from ALB");
        
        return sg;
    }
    
    private Role createInstanceRole() {
        return Role.Builder.create(this, "InstanceRole" + environmentSuffix)
            .assumedBy(ServicePrincipal.Builder.create("ec2.amazonaws.com").build())
            .managedPolicies(List.of(
                ManagedPolicy.fromAwsManagedPolicyName("CloudWatchAgentServerPolicy"),
                ManagedPolicy.fromAwsManagedPolicyName("AmazonSSMManagedInstanceCore")
            ))
            .build();
    }
    
    private LaunchTemplate createLaunchTemplate(SecurityGroup securityGroup, Role instanceRole) {
        InstanceProfile instanceProfile = InstanceProfile.Builder.create(this, "InstanceProfile" + environmentSuffix)
            .role(instanceRole)
            .build();
            
        // User data script to install and configure web server
        String userData = "#!/bin/bash\n" +
            "yum update -y\n" +
            "yum install -y httpd\n" +
            "systemctl start httpd\n" +
            "systemctl enable httpd\n" +
            "echo '<h1>Hello from Web Server</h1>' > /var/www/html/index.html\n" +
            "# Configure CloudWatch agent for log shipping\n" +
            "yum install -y amazon-cloudwatch-agent\n";
            
        return LaunchTemplate.Builder.create(this, "WebAppLaunchTemplate" + environmentSuffix)
            .launchTemplateName("webapp-template-" + environmentSuffix.toLowerCase())
            .machineImage(MachineImage.latestAmazonLinux2023())
            .instanceType(InstanceType.of(InstanceClass.T3, InstanceSize.MICRO))
            .securityGroup(securityGroup)
            .userData(UserData.custom(userData))
            .build();
    }
    
    private AutoScalingGroup createAutoScalingGroup(LaunchTemplate launchTemplate) {
        AutoScalingGroup asg = AutoScalingGroup.Builder.create(this, "WebAppAsg" + environmentSuffix)
            .vpc(vpc)
            .launchTemplate(launchTemplate)
            .minCapacity(2)
            .maxCapacity(6)
            .desiredCapacity(2)
            .vpcSubnets(SubnetSelection.builder()
                .subnetType(SubnetType.PRIVATE_WITH_EGRESS)
                .build())
            .healthCheck(HealthCheck.elb(software.amazon.awscdk.services.autoscaling.ElbHealthCheckOptions.builder()
                .grace(Duration.minutes(5))
                .build()))
            .build();
            
        // Configure CPU-based scaling policies
asg.scaleOnCpuUtilization("CpuScaling" + environmentSuffix,
            software.amazon.awscdk.services.autoscaling.CpuUtilizationScalingProps.builder()
                .targetUtilizationPercent(70)
                .build());
            
        return asg;
    }
    
    private ApplicationLoadBalancer createApplicationLoadBalancer(SecurityGroup securityGroup) {
        return ApplicationLoadBalancer.Builder.create(this, "WebAppAlb" + environmentSuffix)
            .vpc(vpc)
            .internetFacing(true)
            .securityGroup(securityGroup)
            .vpcSubnets(SubnetSelection.builder()
                .subnetType(SubnetType.PUBLIC)
                .build())
            .build();
    }
    
    private void configureSslAndTargetGroups() {
        // Create target group
        ApplicationTargetGroup targetGroup = ApplicationTargetGroup.Builder.create(this, "WebAppTargetGroup" + environmentSuffix)
            .port(80)
            .protocol(ApplicationProtocol.HTTP)
            .targetType(TargetType.INSTANCE)
            .vpc(vpc)
            .build();
            
        // Attach Auto Scaling Group to target group
        autoScalingGroup.attachToApplicationTargetGroup(targetGroup);
        
        // Health check configuration
        targetGroup.configureHealthCheck(software.amazon.awscdk.services.elasticloadbalancingv2.HealthCheck.builder()
            .enabled(true)
            .path("/")
            .protocol(software.amazon.awscdk.services.elasticloadbalancingv2.Protocol.HTTP)
            .healthyHttpCodes("200")
            .interval(Duration.seconds(30))
            .timeout(Duration.seconds(10))
            .healthyThresholdCount(2)
            .unhealthyThresholdCount(3)
            .build());
            
        // Create HTTP listener with target group
        ApplicationListener httpListener = alb.addListener("HttpListener" + environmentSuffix,
            BaseApplicationListenerProps.builder()
                .port(80)
                .protocol(ApplicationProtocol.HTTP)
                .defaultTargetGroups(List.of(targetGroup))
                .build());
    }
    
    private void applyTags() {
        Tags.of(this).add("Environment", "Production");
        Tags.of(this).add("App", "WebApp");
    }
    
    private void createOutputs() {
        CfnOutput.Builder.create(this, "LoadBalancerDns")
            .value(alb.getLoadBalancerDnsName())
            .description("DNS name of the load balancer")
            .exportName("LoadBalancerDNS")
            .build();
            
        CfnOutput.Builder.create(this, "LogsBucketName")
            .value(logsBucket.getBucketName())
            .description("Name of the S3 logs bucket")
            .exportName("S3BucketName")
            .build();
            
        CfnOutput.Builder.create(this, "VpcId")
            .value(vpc.getVpcId())
            .description("VPC ID")
            .exportName("VPCId")
            .build();
            
        CfnOutput.Builder.create(this, "AutoScalingGroupName")
            .value(autoScalingGroup.getAutoScalingGroupName())
            .description("Auto Scaling Group name")
            .exportName("AutoScalingGroupName")
            .build();
    }
    
    // Getters
    public Vpc getVpc() { return vpc; }
    public ApplicationLoadBalancer getApplicationLoadBalancer() { return alb; }
    public AutoScalingGroup getAutoScalingGroup() { return autoScalingGroup; }
    public Bucket getLogsBucket() { return logsBucket; }
}